File: LeetCode/212-word-search-ii.py
###############################################################################################
# 暴搜直接超时，改成trie后时间复杂度还是高
###########
# 时间复杂度：O(m*n*(3^l)), l 是最长单词的长度
# 空间复杂度：O(k*l), k是单词数，l是最长单词长度，用于存sons
###############################################################################################
class Solution:
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        m, n = len(board), len(board[0])
        sons = [defaultdict(int) for _ in range(2)]
        isword = [0]*2
        idx = 1
        
        def insert(word):
            nonlocal idx
            p = 0
            for i in range(len(word)):
                id_ = ord(word[i]) - ord('a')
                if id_ not in sons[p]:
                    sons[p][id_] = idx
                    idx += 1
                    sons.append(defaultdict(int))
                    isword.append(0)
                p = sons[p][id_]
            isword[p] = 1
        
        for word in words:
            insert(word)

        dirs = [[-1, 0], [0, 1], [1, 0], [0, -1]]
        def dfs(x, y, p, path):
            if ord(board[x][y]) - ord('a') not in sons[p]:
                return
            p = sons[p][ord(board[x][y]) - ord('a')]
            path += board[x][y]
            if isword[p]:
                res.append(path)
                isword[p] = 0 # 只加入一次
            
            for ux, uy in dirs:
                nx, ny = x+ux, y+uy
                if (nx, ny) not in st and 0 <= nx < m and 0 <= ny < n:
                    st.add((nx, ny))
                    dfs(nx, ny, p, path)
                    st.remove((nx,ny))

        res = []
        for i in range(m):
            for j in range(n):
                st = set()
                st.add((i,j))
                dfs(i, j, 0, "")
        return res

File: LeetCode/212-word-search-ii.cpp
///////////////////////////////////////////////////////////////////////////////////////////////
// c++ version trie, but this TLE
////////////////
// Time Complexity: O(n_words + m*n*(4**10))
// Space Complexity: O(n_words)
///////////////////////////////////////////////////////////////////////////////////////////////
class TrieNode{
public:
    bool is_end;
    vector<TrieNode*> children;
    TrieNode(){
        is_end=false;
        children=vector<TrieNode*>(26, NULL);
    }   
};

class Trie{
public:
    TrieNode* getRoot(){return root;}
    Trie(vector<string>& words){
        root=new TrieNode();
        for(int i=0; i<words.size(); ++i)
            addWord(words[i]);
    }
    void addWord(const string& word){
        TrieNode* cur=root;
        for(int i=0; i<word.size(); ++i){
            int index=word[i]-'a';
            if(cur->children[index]==NULL)   
               cur->children[index]=new TrieNode();
            cur=cur->children[index];    
        }
        cur->is_end=true;
    }
private:
    TrieNode* root;
};

class Solution {
public:
    vector<string> findWords(vector<vector<char>>& board, vector<string>& words) {
        Trie* trie = new Trie(words);
        TrieNode* root=trie->getRoot();
        set<string> result_set;
        for(int x=0; x<board.size(); ++x)
            for(int y=0; y<board[0].size(); ++y)
                dfs(board, x, y, root, "", result_set);
        
        vector<string> res;
        for(auto it:result_set)    res.push_back(it);
        return res;
    }
    void dfs(vector<vector<char>>& board, int x, int y, TrieNode* root, string word, set<string>& result){
        // use ' ' to record walked cell
        if(x<0||x>=board.size()||y<0||y>=board[0].size() || board[x][y]==' ') return;
        
        // only go when this word exists in trie
        if(root->children[board[x][y]-'a'] != NULL){
            word=word+board[x][y];
            root=root->children[board[x][y]-'a']; 
            if(root->is_end) result.insert(word);
            
            char c=board[x][y]; // record for recovering
            board[x][y]=' ';
            dfs(board, x+1, y, root, word, result);
            dfs(board, x-1, y, root, word, result);
            dfs(board, x, y+1, root, word, result);
            dfs(board, x, y-1, root, word, result);
            board[x][y]=c;
        }
    }
};